Support local certificates and HTTP proxies

mariovisic opened the pull request:
Hi there 👋 

Found your library as we need to check mTLS certificates using OCSP, thanks for the hard work figuring this all out.

There's a couple of features that I need for our environment:

- We need to specify the proxy host and port for the OCSP checks, but not have them apply globally (so we can't set an environment variable for all net/http calls).
- We already have the client certificate to check as well as the CA bundle, so rather than making a call to a URL to retrieve the chain, instead we can grab the chain from our existing certificates

Here's a patch that implements the changes, I haven't yet included any documentation updates or tests, just keen to firstly hear if you'd be interested in accepting the changes and any feedback.

I also haven't changed the API at all, but it may make sense to add a new class method so that `test` doesn't have two different ways of retrieving the certificate and chain, setting `url` to `nil` feels a bit odd.

Here's the usage when setting the certs and proxy!

```ruby
client_cert = OpenSSL::X509::Certificate.new(File.read("bad_cert.cer"))
ca_certs = OpenSSL::X509::Certificate.load(File.read("ca-bundle-cts.cer"))

SSLTest.test(nil, client_cert: , ca_certs: ca_certs, proxy_host: '...', proxy_port: '...')
```

mariovisic committed:
Support local certificates and HTTP proxies

- Allow specifying the local certificate to check and the CA bundle rather
than looking up the certificates via URL.
- Support providing a proxy host and port

jarthod commented:
Hi @mariovisic :wave: 

Thank you for your contribution, I am interested in accepting those changes after a few updates to the API (and test/readme updates).
1. The proxy change looks ok to me, and you also added it to the CRL method which is good. So this one would just need some tests and documentation in the readme.
2. The API for local cert testing as you said doesn't look great this way and would definitely need two methods I think (something like `#test_cert` and `#test_url` maybe?, `#test` could be an alias for `#test_url` for backward compatibility although this gem doesn't have many users beside me so not a big deal ^^)
It doesn't look like a lot of code would be shared between both methods anyway, a little bit of duplication is fine to me if it makes reading the code simpler, if there's a substantial portion of code to share you can extract it of course.

While we are doing this, and as you seem to understand client and CA certs, I'm not sure how complicated it would be to add the possibility to pass some `ca_certs` to the `#test_url` method in order to validate the server (downloaded) certificate against a private CA, I might use that feature later for https://updown.io ^^ But don't spend your time on this unless you know how to do it and it's easy.

Thanks!

mariovisic commented:
Neat, i'm excited to get the changes in, I can start working on docs and tests. When I run the current test suite, quite a few tests are failing, i'm assuming because the live data has changed, so I can try send a patch for getting that all green first before merging this PR.

I really like the idea of `test_cert` and `test_url` are you happy for those to be class methods just like `test` is now?

> and as you seem to understand client and CA certs

😬  oh ... I wouldn't say that 😅 


> I'm not sure how complicated it would be to add the possibility to pass some ca_certs to the #test_url method in order to validate the server (downloaded) certificate against a private CA

That would be really straightforward I believe just using the equals operator `==`, from the docs:
https://ruby-doc.org/3.1.3/exts/openssl/OpenSSL/X509/Certificate.html#method-c-load

> Compares the two certificates. Note that this takes into account all fields, not just the issuer name and the serial number.

If you want to compare only some of the fields are the same then we could do that specifically!

If you have some test certificates and a URL then I can verify it works as expected 👍

jarthod commented:
Thanks for the test update :)
Yes I'm fine with `test_cert` and `test_url` being class methods just like `test `.

I'm not sure it's as simple as comparing certificates, the server certificate will not be "equal" to the private CA cert, it'll be "signed" by the private CA cert. It's another link in the chain, instead of verifying who signed the server certificate from a list of trusted CA installed in the OS, we would check if it was signed by one of the passed private CA cert. But in any case don't worry about this too much and just do what you planned ^^ If I find (or create) a good example I'll see about that later.

mariovisic commented:
Ooooh, right! I believe you would do the same thing that is in my patch, create a x509 store, then add the CA certificates, then call `store.verify(client_cert)`. It'll error if the `client_ceret` isn't signed by one of the CA certificates added to the store. By default it does not use the CA certificates on the system, you need to call: `#set_default_paths` to use the system ca certs: https://ruby-doc.org/stdlib-3.0.0/libdoc/openssl/rdoc/OpenSSL/X509/Store.html#method-i-set_default_paths

mariovisic committed:
Update method names to be test_url and test_cert

Extract the logic of testing a cert to its own method and add new
methods for common functionality!

I've added one test for the new functionality, will start to add more
now!

mariovisic commented:
I've separated the two methods and added the alias, I think the only thing left to do is to add test coverage and documentation and we're good. There are some things that `test_cert` won't do, for example self signed certificates won't be picked up because you're supplying the CA certs, it's just checking against what you provide.

Let me know if anything sticks out that needs changing, otherwise i'll just plug away at the tests, should be good by tomorrow I reckon 👌

jarthod commented on lib/ssl-test.rb:
```diff
@@ -13,13 +13,14 @@ module SSLTest
 
   class << self
-    def test url, open_timeout: 5, read_timeout: 5, redirection_limit: 5
+    def test_url url, open_timeout: 5, read_timeout: 5, proxy_host:nil, proxy_port: nil, redirection_limit: 5, ca_certs: []
```
The `ca_certs` parameter is not used here, it's here for the eventual other feature?

mariovisic replied:
This was for the feature that you were describing above, but I can do that in a followup PR 👍

jarthod commented on lib/ssl-test/ocsp.rb:
```diff
@@ -82,7 +82,7 @@ def follow_ocsp_redirects(uri, data, open_timeout: 5, read_timeout: 5, redirecti
         [http_response.body, nil]
       when Net::HTTPRedirection
-        follow_ocsp_redirects(URI(http_response["location"]), data, open_timeout: open_timeout, read_timeout: read_timeout, redirection_limit: redirection_limit - 1)
+        follow_ocsp_redirects(URI(http_response["location"]), data, open_timeout: open_timeout, read_timeout: read_timeout, rproxy_host: proxy_host, proxy_port: proxy_port, edirection_limit: redirection_limit - 1)
```
Some typos here (rproxy_host, ederection_limit): if you can, make sure the specs would catch those ^^

mariovisic replied:
Neat, have fixed and added proxy tests which cover this 👍

jarthod commented:
Thanks! I just left 2 small comments. Yes for self-signed certificates it makes sense, all custom CA certs can be qualified as "self-signed" as long as you don't have the parent CA ^^ (and if you have it they are valid).

mariovisic committed:
Add tests and fixes for new test_cert method

Changes include:
  - Copied over 4 tests from the test_url method
  - Fixed a typo in redirection_limit (with test coverage)
  - Fixed a typo in proxy_port (will add test coverage for this soon)

mariovisic committed:
Add tests for proxy config

mariovisic committed:
Implement remaining tests for test_cert method

mariovisic committed:
Update documentation to include the new test_cert method!

mariovisic commented:
Thanks for being patient, this is ready for review now, I've updated all the tests and the documentation on the readme!

jarthod commented on spec/ssl-test_spec.rb:
```diff
@@ -8,7 +10,16 @@
+    if proxy_thread
+      proxy_thread.kill
+    end
+  end
```
I would probably shorten this as:
```suggestion
  after(:each) { proxy_thread&.kill }
```

mariovisic replied:
Great, thank you :)

jarthod commented on spec/ssl-test_spec.rb:
```diff
@@ -235,4 +277,189 @@
+          cert = OpenSSL::X509::Certificate.new(File.read(File.join(__dir__, 'fixtures/google_com_client.pem')))
+          ca_bundle = OpenSSL::X509::Certificate.load(File.read(File.join(__dir__, 'fixtures/google_com_ca_bundle.pem')))
+
+          valid, error, cert = SSLTest.test_cert(cert, ca_bundle, proxy_host: '127.0.0.1', proxy_port: 8080)
```
I'm not sure how to change it, but this is actually not testing that we use the proxy, if we remove all the code using "proxy_host" and "proxy_port" and just leave the unused parameters, this test would still pass.
Can you check if the WEBrick::HTTPProxyServer has something which changes on request (a counter or something) that we could use to make sure ? Otherwise I guess you can use the logger or access-log (in memory like with a StringIO or something) to verify the expected request were all proxied ?
(some thing above in the `test_url` specs of course)

mariovisic replied:
That is true, I think combined with the following test not being able to connect, it does indirectly show the proxy is working, but you're right, if I can spy on the requests to the proxy (in theory this should work fine, but unsure if the thread will get in the way), then we can verify a request has gone to the proxy 👍

mariovisic replied:
Updated in 46dd72d389ad27b2c3f0ac6919463cf50622f366

jarthod commented on lib/ssl-test.rb:
```diff
@@ -81,6 +105,28 @@ def logger= logger
 
     private
 
+    def parsed_message(revoked, revocation_date, message)
```
Just a note about naming here, this function name is not explicit enough IMO, maybe something like `revocation_message` ? (we're not really parasing anything here, just generating the revocation error message if any)

mariovisic replied:
Good suggestion, thank you, updated in 2ac80b42188c33f995ca2245954c9d8760fbbf07

jarthod reviewed (commented):
Pretty good thanks ! I left a few comments.
I did not expect this many tests and fixtures (I see you copied all the cases from the URL check), not sure if they are all relevant to a client cert test but that's fine for me.

mariovisic committed:
Update spec/ssl-test_spec.rb

Co-authored-by: Adrien Rey-Jarthon <jobs@adrienjarthon.com>

mariovisic committed:
Rename parsed_message to revocation_message

mariovisic commented:
> Pretty good thanks ! I left a few comments. I did not expect this many tests and fixtures (I see you copied all the cases from the URL check), not sure if they are all relevant to a client cert test but that's fine for me.

Ah right, I thought I'd only copied the relevant ones, I'm happy to remove tests that aren't testing anything useful

mariovisic committed:
Update test to ensure the proxy receives requests :ok_hand:

mariovisic commented:
I've updated the tests to check that the proxy received requests 👌  and it's all working well, checkout commit: 46dd72d389ad27b2c3f0ac6919463cf50622f366

I think this is ready for review again :)

jarthod commented on spec/ssl-test_spec.rb:
```diff
@@ -176,6 +183,41 @@
+          proxy_thread
+          sleep 0.1 # wait for the proxy to start!
+
+          allow($proxy).to receive(:do_GET).and_call_original
```
Thanks for the assertion, quick note: I believe you can do it in a single line with:
```suggestion
          expect($proxy).to receive(:do_GET).twice.and_call_original
```
(no need for the `have_received` in this case) but to be verified ^^

mariovisic replied:
Yeah totally, that will work great, I usually avoid using the mock expectation that `expect().to receive` creates  because the test structure is harder to follow ... eg: with the existing code it's `setup -> execution -> expectation`, which is the same pattern for tests that aren't isolating with stubs/mocks/spies etc...  where as using the mock the steps are: `setup -> expectation -> execution` which is harder to follow as it's different to other tests:


Here's a test without any external dependencies
```ruby
# setup (this could be in a let as well)
user = User.new

# execution
result = user.name

# expectation
expect(result).to eq('Adrien')
```

Here's an example using a spy via `allow`
```ruby
# setup
user = User.new
allow(MyExternalDependency).to receive(:test)

# execution
user.name

# expectation
expect(MyExternalDependency) to have_received(:test)
```

Here's the same with the mock example via `expect ... to receive`
```ruby
# setup
user = User.new

# expectation
expect(MyExternalDependency).to receive(:test)

# execution
user.name
```

mariovisic replied:
I don't have a strong opinion on one or the other, just wanted to go through my rationale, happy to change it 👍

jarthod replied:
Indeed, I am used to the having the expectation first with rspec mocks so that does not bother me much but I understand the rational and I agree it's easier to read/understand the way you wrote it. Let's keep it that way!

In that same spirit though, wouldn't it be a good idea to move the `allow($proxy).to receive(:do_GET).and_call_original` line inside the proxy setup `let` ? (as it's not serving any purpose for understanding of the test, it's just required to setup rspec mock so we can count the number of call at the end)

mariovisic replied:
Good idea, added in 137994a3225663d627d077eb9f50dd20fc0d5005  I moved the `sleep` too as it's part of setting up the proxy!

jarthod replied:
Nice! and sorry but I just noticed the "let(:proxy_thread)" block is present twice and identical? can't we move it up to the top block?

mariovisic replied:
Yes, though will need to solve the problem of killing it once it's done. We have this code: `proxy_thread&.kill` to stop the proxy which will boot the proxy when calling `proxy_thread` if we moved `proxy_thread` to the top.

We could boot the proxy once for all of the tests, though we'd need to reset the stub each time.

killing the proxy could be moved to the test body but it won't run if the test runs into an exception.

mariovisic replied:
... it'll boot the proxy as the `let` is lazily loaded

jarthod replied:
Ah yep ok, let's keep it that way then :)

jarthod reviewed (approved):
Thanks, looks good to me 👍 
I let you have a look at the final comment and I'll merge after that.

mariovisic committed:
Update formatting of specs for proxy thread

Move the stubbing and sleep required for proxy setup to the `let` out of
the test body for readability.

jarthod commented on spec/ssl-test_spec.rb:
```diff
@@ -176,6 +183,43 @@
+          expect(valid).to be_nil
+          expect(cert).to be_nil
+        end
+      end
```
But if I'm correct, it means this test will actually also start the proxy during the teardown (due to `after(:each) { proxy_thread&.kill }`) while not needing it at all during the test? Not a big deal but maybe the proxy should be in the `context 'when the proxy is active' do` then?

jarthod commented on spec/ssl-test_spec.rb:
```diff
@@ -235,4 +279,195 @@
+          expect(valid).to be_nil
+          expect(cert).to eq(cert)
+        end
+      end
```
Same thing here of course

mariovisic commented:
good catch! Totally. You’re right, we should move the let to be inside the specific context. I’m away from my PC so can do this with tonight, I’m also happy if you want to commit the change if that’s easier!On 22 Nov 2025, at 10:20 pm, Adrien Rey-Jarthon ***@***.***> wrote:﻿
@jarthod commented on this pull request.



In spec/ssl-test_spec.rb:
> +
+          expect($proxy).to have_received(:do_GET).once
+        end
+      end
+
+      context 'when the proxy is not reachable' do
+        it 'returns a http error' do
+          cert = OpenSSL::X509::Certificate.new(File.read(File.join(__dir__, 'fixtures/google_com_client.pem')))
+          ca_bundle = OpenSSL::X509::Certificate.load(File.read(File.join(__dir__, 'fixtures/google_com_ca_bundle.pem')))
+
+          valid, error, cert = SSLTest.test_cert(cert, ca_bundle, proxy_host: '127.0.0.1', proxy_port: 55000)
+          expect(error).to include('(Connection refused - connect(2) for "127.0.0.1" port 55000)')
+          expect(valid).to be_nil
+          expect(cert).to eq(cert)
+        end
+      end

Same thing here of course

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

jarthod commented:
I just merged and changed the spec in the merge commit :)
I also released `1.5.0` with this change.

mariovisic commented:
> I just merged and changed the spec in the merge commit :) I also released `1.5.0` with this change.

Thank you :) what great news, thanks for collaborating to get this over the line!
